test/inflightRequests.js

```diff
@@ -0,0 +1,205 @@
+            assert.strictEqual(HTTPCLIENT.inflightRequests(), 1);
+
+            req.on('result', function (_err, res) {
+                assert.ifError(_err);
+                res.on('data', _.noop);
+                res.on('end', function () {
+                    assert.strictEqual(HTTPCLIENT.inflightRequests(), 0);
+                    return done();
+                });
+            });
+        });
+    });
+
+    it('HttpClient should increment and decrement connect timeout',
+    function (done) {
+        // setup client to point to unresolvable IP
+        var client = clients.createHttpClient({
+            url: 'http://10.255.255.1/',
+            connectTimeout: 100,
+            retry: false
+        });
+
+        client.get('/timeout', function (err, req) {
+            assert.ok(err);
+            assert.strictEqual(err.name, 'ConnectTimeoutError');
+            assert.strictEqual(client.inflightRequests(), 0);
+            return done();
+        });
+
+        assert.strictEqual(client.inflightRequests(), 1);
+    });
+
+    it('HttpClient should increment and decrement request timeout',
+    function (done) {
+        HTTPCLIENT.get('/timeout', function (err, req) {
+            // no connect timeout
+            assert.ifError(err);
+            assert.strictEqual(HTTPCLIENT.inflightRequests(), 1);
+
+            req.on('result', function (_err, res) {
+                assert.ok(_err);
+                assert.strictEqual(_err.name, 'RequestTimeoutError');
+                assert.notOk(res);
+                assert.strictEqual(HTTPCLIENT.inflightRequests(), 0);
+                return done();
+            });
+        });
+    });
+
+    it('HttpClient should increment and decrement on forced req abort',
+    function (done) {
+
+        var client = clients.createHttpClient({
+            url: 'http://localhost:3000'
+        });
+
+        client.get('/timeout', function (err, req) {
+            // no connect timeout
+            assert.ifError(err);
+            assert.strictEqual(client.inflightRequests(), 1);
+
+            req.on('result', function (_err, res) {
+                assert.strictEqual(client.inflightRequests(), 0);
+                assert.ok(_err);
+                assert.ok(_err.message, 'socket hang up');
+                assert.isNull(res);
+                return done();
+            });
+
+            req.abort();
+        });
+    });
+
+    it('should count multiple inflight requests', function (done) {
+
+        STRINGCLIENT.get('/200', _.noop);
+        assert.strictEqual(STRINGCLIENT.inflightRequests(), 1);
+
+        STRINGCLIENT.get('/200', _.noop);
+        assert.strictEqual(STRINGCLIENT.inflightRequests(), 2);
+
+        STRINGCLIENT.get('/200', _.noop);
+        assert.strictEqual(STRINGCLIENT.inflightRequests(), 3);
+
+        STRINGCLIENT.get('/200', _.noop);
+        assert.strictEqual(STRINGCLIENT.inflightRequests(), 4);
+
+        setTimeout(function () {
```

Comment:
On a busy system, it could timeout _before_ all requests and responseshave been processed, and thus could make this test flaky.

What do you think of passing the same callback to all `STRINGCLIENT.get` calls that would decrement a counter initialized at 4, close the server when that counter reaches 0, and then call `done()` in the listener for the server's `'close'` event?

Reply:
Good call. I was being lazy. 👍